File: nova/registro.py
# -*- coding: utf-8 -*-
"""Il registro delle azioni che non si annullano.

Perche' esiste. NOVA gira in autonomia piena e da qualche giorno manda
candidature di lavoro al posto dell'utente. Un foglio sbagliato si rifa', una
mail sbagliata si corregge con un'altra mail; una domanda di lavoro parte,
arriva a una persona che non conosci, e ti giudica.

Questo non e' un freno, ed e' una scelta esplicita: N1 dice che NOVA non ha
confini e N9 che il confine e' una manopola dell'utente. Mettere un cancello
che scavalca la richiesta sarebbe tradire la premessa per far stare tranquillo
chi ha scritto il codice.

Ma la responsabilita' ha bisogno di **visibilita'**: si risponde solo di
quello che si puo' vedere. Se NOVA manda tre candidature mentre l'utente
guarda altrove, senza registro non resta traccia di cosa e' partito e a chi -
e la responsabilita' resta teorica. Questo file la rende esercitabile. E'
N8 (nessuna morte silenziosa) e N10 (ogni azione dichiara cio' che costa),
non un permesso da chiedere.

Due sorgenti, di proposito:

- **automatica**, dentro gli strumenti che cambiano il mondo attraverso il
  browser. Il modello puo' dimenticarsi di annotare; una riga scritta dallo
  strumento no;
- **esplicita**, con `azione_registra`, per tutto il resto - una mail
  inviata, una candidatura, un acquisto - dove solo NOVA sa che quel click
  era il punto di non ritorno.

Il valore di una credenziale non entra qui, mai. Ne entra il nome (N4).
"""
from __future__ import annotations

import json
import os
import time
from datetime import datetime
from pathlib import Path

from .forme_riservate import etichetta_di_segreto, maschera

# Oltre questa soglia il file viene ruotato in .1: un registro che cresce
# senza fine e' un registro che nessuno apre.
from .rotazione import MAX_BYTE as BYTE_MAX          # noqa: F401
TESTO_MAX = 300


def percorso() -> Path:
    base = os.environ.get("APPDATA")
    radice = Path(base) / "NOVA" if base else Path.home() / ".config" / "NOVA"
    return radice / "azioni.jsonl"


def _ruota(f: Path) -> None:
    """Due megabyte, poi si ricomincia e il precedente resta.

    Faceva questa cosa per conto suo, con un tetto suo: 2.000.000 di byte
    invece di 2.097.152, cioe' due megabyte da fruttivendolo invece che due
    megabyte veri. Nessuno l'avrebbe mai notato — e' proprio per questo che
    adesso il tetto sta in un posto solo (D72).
    """
    from .rotazione import ruota_se_serve
    ruota_se_serve(f, BYTE_MAX)


def _dettagli_sicuri(azione: str, dove: str, dettagli: str) -> str:
    """I dettagli mascherati, e sostituiti del tutto se il campo li annuncia.

    Il caso che il filtro per forme non puo' prendere: NOVA compila un modulo
    di accesso e scrive «scritto in #password» nell'azione e «Tramonto2026!»
    nei dettagli. Guardati uno per volta non sono niente — la seconda e' una
    parola con dentro un anno. Guardati insieme sono una credenziale.

    Saper compilare un modulo di accesso e' una cosa che NOVA deve fare. Il
    prezzo e' che il registro di quelle azioni non puo' conservare cio' che ha
    scritto: resta la riga, che dice cosa e' successo e dove, e sparisce il
    valore, che e' l'unica parte che non serve a nessuno per rileggere la
    storia.
    """
    if etichetta_di_segreto(azione) or etichetta_di_segreto(dove):
        return "[non registrato: il campo contiene una credenziale]"
    return maschera(dettagli or "")[:TESTO_MAX]


def annota(azione: str, dove: str = "", dettagli: str = "",
           tipo: str = "browser", esito: str = "") -> None:
    """Scrive una riga. Non solleva mai: un registro che impedisce di
    lavorare verrebbe tolto di mezzo dopo mezza giornata, ed e' peggio che
    non averlo."""
    try:
        f = percorso()
        f.parent.mkdir(parents=True, exist_ok=True)
        _ruota(f)
        # Tutto quello che entra qui passa dal filtro, senza eccezioni.
        #
        # Il registro e' un file che resta, e ci finisce dentro anche il testo
        # che NOVA **scrive** nei campi: `annota("scritto in ...",
        # dettagli=testo)`. NOVA sa compilare un modulo di accesso — e' una
        # cosa che deve saper fare — quindi prima o poi in `dettagli` c'e' una
        # password. Ci finiscono anche le righe di comando, e una riga di
        # comando porta volentieri un `Authorization: Bearer`.
        #
        # Si maschera **qui** e non nei quindici posti che chiamano `annota`,
        # per la stessa ragione per cui il vault si chiude su `upsert`: la
        # porta e' una sola, e chiuderla li' vuol dire chiuderla e basta. Un
        # chiamante che si dimentica non e' un'ipotesi, e' una certezza.
        riga = {
            "quando": datetime.now().isoformat(timespec="seconds"),
            "tipo": tipo,
            "azione": maschera(azione or "")[:200],
            "dove": maschera(dove or "")[:300],
            "dettagli": _dettagli_sicuri(azione, dove, dettagli),
        }
        if esito:
            riga["esito"] = maschera(esito)[:200]
        with open(f, "a", encoding="utf-8") as fh:
            fh.write(json.dumps(riga, ensure_ascii=False) + "\n")
    except Exception:
        pass


def leggi(quante: int = 30, ore: float = 0) -> list[dict]:
    """Le ultime righe, dalla piu' recente. Con `ore` si guarda una finestra.

    Legge **anche lo storico**, e prima di quello vivo. Il registro si pota,
    e questo e' l'unico file potato di NOVA su cui qualcuno fa una domanda
    vecchia: `cerca` esiste per «cosa ho mandato a quella societa'?» tre
    settimane dopo. Leggendo solo il file vivo, il giorno della potatura
    quella domanda avrebbe cominciato a rispondere «niente» — senza errori,
    senza righe di log, e senza che si potesse capire perche'.
    """
    f = percorso()
    vecchio = f.with_suffix(".1" + f.suffix)
    righe: list[dict] = []
    for parte in (vecchio, f):        # prima il vecchio: l'ordine e' il tempo
        if not parte.exists():
            continue
        try:
            with open(parte, encoding="utf-8") as fh:
                for r in fh:
                    r = r.strip()
                    if not r:
                        continue
                    try:
                        righe.append(json.loads(r))
                    except Exception:
                        continue
        except Exception:
            continue
    if ore:
        limite = time.time() - ore * 3600
        tenute = []
        for x in righe:
            try:
                if datetime.fromisoformat(x["quando"]).timestamp() >= limite:
                    tenute.append(x)
            except Exception:
                tenute.append(x)
        righe = tenute
    return righe[-quante:][::-1]


def _senza_accenti(s: str) -> str:
    import unicodedata
    return "".join(c for c in unicodedata.normalize("NFD", s or "")
                   if unicodedata.category(c) != "Mn").casefold()


def cerca(testo: str = "", tipo: str = "", esito: str = "",
          giorni: float = 0, quante: int = 50) -> list[dict]:
    """«Cosa ho mandato a quella societa'?», tre settimane dopo.

    Un registro che si puo' solo scorrere dalla fine e' un registro che si
    legge il primo giorno. La domanda vera arriva dopo, ed e' sempre della
    stessa forma: una parola che ci si ricorda, e un periodo vago.

    Le parole si cercano tutte, in qualunque campo e in qualunque ordine, e
    senza accenti: chi cerca «societa» deve trovare «societa'», e chi scrive
    di fretta non mette le maiuscole.
    """
    righe = leggi(quante=10_000, ore=giorni * 24 if giorni else 0)
    parole = [_senza_accenti(x) for x in (testo or "").split() if x]
    fuori = []
    for r in righe:
        if tipo and r.get("tipo") != tipo:
            continue
        if esito and esito not in (r.get("esito") or ""):
            continue
        if parole:
            dentro = _senza_accenti(" ".join(
                str(r.get(k, "")) for k in ("azione", "dove", "dettagli", "esito", "tipo")))
            if not all(w in dentro for w in parole):
                continue
        fuori.append(r)
    return fuori[:quante]


def riassunto() -> str:
    """Quanto c'e' dentro, di che tipo, da quando, e dove sta il file.

    E' la risposta a «che cos'e' questo registro»: senza, l'unica strada per
    saperlo e' aprire un .jsonl, e a quel punto non lo apre nessuno.
    """
    righe = leggi(quante=100_000)
    f = percorso()
    if not righe:
        return f"Il registro e' vuoto.\nSta in {f}"
    tipi: dict[str, int] = {}
    for r in righe:
        tipi[r.get("tipo") or "?"] = tipi.get(r.get("tipo") or "?", 0) + 1
    ordinati = sorted(tipi.items(), key=lambda x: -x[1])
    prima = (righe[-1].get("quando") or "")[:10]
    ultima = (righe[0].get("quando") or "")[:10]
    quando = f"dal {_data_italiana(prima)}" if prima == ultima else \
        f"dal {_data_italiana(prima)} al {_data_italiana(ultima)}"
    return (f"{len(righe)} azioni registrate, {quando}.\n"
            + "  " + ", ".join(f"{n} {k}" for k, n in ordinati)
            + f"\nSta in {f}")


def _data_italiana(iso: str) -> str:
    try:
        a, m, g = iso.split("-")
        return f"{g}/{m}/{a}"
    except Exception:                                       # noqa: BLE001
        return iso


def _giorno(iso: str) -> str:
    """«oggi», «ieri», oppure la data. Un timestamp ISO non e' un giorno."""
    from datetime import date, timedelta
    try:
        q = date.fromisoformat(iso[:10])
    except Exception:                                       # noqa: BLE001
        return iso[:10]
    oggi = date.today()
    if q == oggi:
        return "oggi"
    if q == oggi - timedelta(days=1):
        return "ieri"
    return _data_italiana(iso[:10])


def racconta(quante: int = 30, ore: float = 0, righe: list[dict] | None = None) -> str:
    """Le stesse righe, in una forma che si legge senza decodificare JSON.

    Raggruppate per giorno, perche' la domanda a cui questo risponde e'
    «cosa hai fatto ieri» e non «cosa hai fatto alla riga 47».
    """
    righe = leggi(quante, ore) if righe is None else righe
    if not righe:
        return "Nessuna azione registrata."
    fuori = [f"{len(righe)} azioni, dalla piu' recente:"]
    giorno_scritto = ""
    for x in righe:
        quando = x.get("quando") or ""
        giorno = _giorno(quando)
        if giorno != giorno_scritto:
            fuori.append(f"\n— {giorno} —")
            giorno_scritto = giorno
        ora = quando[11:16]
        pezzi = [f"{ora}  [{x.get('tipo')}]  {x.get('azione')}"]
        if x.get("dove"):
            pezzi.append(f"          su: {x['dove']}")
        if x.get("dettagli"):
            pezzi.append(f"          {x['dettagli']}")
        if x.get("esito"):
            pezzi.append(f"          esito: {x['esito']}")
        fuori.append("\n".join(pezzi))
    return "\n".join(fuori)

File: core/crates/nova-core/src/registro.rs
//! Il registro delle azioni che non si annullano, dal lato del demone.
//!
//! NOVA ne aveva gia' uno, in Python, e diceva di se' la cosa giusta: «la
//! responsabilita' ha bisogno di visibilita', si risponde solo di quello che
//! si puo' vedere». Ci finivano le candidature inviate, i click su un
//! pulsante, i moduli compilati.
//!
//! Non ci finiva niente di quello che fa **il demone**, che e' il pezzo di
//! NOVA che gira quando non c'e' nessuno a guardare: `shell.exec` esegue un
//! comando qualunque nella shell del sistema, `fs.write` riscrive un file.
//! Il demone teneva il suo giornale — quello serve ad **annullare**, ed e'
//! un'altra domanda — e chi chiedeva a NOVA «cosa hai fatto?» si sentiva
//! rispondere con meta' della storia, senza sapere che era meta'.
//!
//! Qui c'e' l'altra meta', e scrive **lo stesso file**: due registri sarebbero
//! due risposte alla stessa domanda. La forma della riga — il mascheramento,
//! i tagli, cosa si perde quando il campo annuncia una credenziale — sta in
//! `nova-registro`, con il Python di fronte e un banco che li confronta
//! carattere per carattere.

use std::io::Write;
use std::path::{Path, PathBuf};

use nova_registro::{Filtro, Riga, Scritta};

/// Quante righe si leggono quando nessuno dice quante.
pub const QUANTE: usize = 30;

/// Dove sta il registro: lo stesso file del Python, e non per caso.
pub fn percorso() -> PathBuf {
    let base = std::env::var("APPDATA")
        .map(PathBuf::from)
        .or_else(|_| std::env::var("HOME").map(|h| PathBuf::from(h).join(".config")))
        .unwrap_or_else(|_| std::env::temp_dir());
    base.join("NOVA").join("azioni.jsonl")
}

/// L'ora come la scrive il Python: `2026-08-30T14:18:05`, **locale**.
///
/// Locale e non UTC, e non e' formattazione: su questo file scrivono tutte e
/// due le meta' di NOVA, e la parte Python scrive l'ora dell'orologio di chi
/// sta davanti al computer. Due righe della stessa giornata sfalsate di
/// un'ora non darebbero nessun errore — sarebbero solo sbagliate, e chi
/// rilegge non ha modo di accorgersene.
pub(crate) fn adesso() -> String {
    let t = nova_platform::orologio::adesso();
    quando(t, nova_platform::fuso_secondi(t))
}

/// Quell'istante, in quel fuso, come lo scrive il Python.
///
/// Separata da [`adesso`] perche' cosi' si puo' provare: con l'orologio e il
/// fuso dentro, l'unica prova possibile sarebbe rifare lo stesso conto — e su
/// una macchina regolata su UTC, che e' quella della CI, non distinguerebbe
/// **niente** da un demone che scrive l'ora di Greenwich.
fn quando(istante: i64, fuso_secondi: i64) -> String {
    nova_calendario::da_istante(istante, fuso_secondi).iso()
}

/// Scrive una riga. **Non fallisce mai**, come dall'altra parte: un registro
/// che impedisce di lavorare viene tolto di mezzo dopo mezza giornata, ed e'
/// peggio che non averlo.
pub fn annota(azione: &str, dove: &str, dettagli: &str, tipo: &str, esito: &str) {
    annota_in(&percorso(), azione, dove, dettagli, tipo, esito);
}

/// Come [`annota`], ma su un file scelto.
///
/// Il file si passa invece di leggerlo da una variabile d'ambiente: una
/// variabile e' del **processo**, e due prove che la cambiano insieme si
/// pestano i piedi — cosa che e' successa subito, la prima volta che ho
/// scritto queste prove.
pub fn annota_in(f: &Path, azione: &str, dove: &str, dettagli: &str, tipo: &str, esito: &str) {
    let s = Scritta {
        azione,
        dove,
        dettagli,
        tipo,
        esito,
    };
    let riga = nova_registro::riga_da_scrivere(&s, &adesso());
    if let Some(dir) = f.parent() {
        let _ = std::fs::create_dir_all(dir);
    }
    // Si pota **prima** di scrivere: dopo vuol dire che il file supera sempre
    // il tetto di una riga, e su un file che si apre per capire cos'e'
    // successo l'ultima riga e' quella che conta.
    nova_potatura::ruota_se_serve(f);
    if let Ok(mut fh) = std::fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(f)
    {
        let _ = writeln!(fh, "{riga}");
    }
}

/// Le ultime righe, dalla piu' recente.
///
/// Legge **anche lo storico**, e prima di quello vivo. Il registro si pota, ed
/// e' l'unico file potato di NOVA su cui si fa una domanda vecchia: «cosa ho
/// mandato a quella societa'?» tre settimane dopo. Leggendo solo il file vivo,
/// il giorno della potatura quella domanda comincerebbe a rispondere
/// «niente» — senza errori e senza che si possa capire perche'.
pub fn leggi(quante: usize) -> Vec<Riga> {
    leggi_da(&percorso(), quante)
}

/// Come [`leggi`], ma da un file scelto.
pub fn leggi_da(f: &Path, quante: usize) -> Vec<Riga> {
    let mut righe: Vec<Riga> = Vec::new();
    for parte in [nova_potatura::precedente(f), f.to_path_buf()] {
        let Ok(testo) = std::fs::read_to_string(&parte) else {
            continue;
        };
        righe.extend(testo.lines().filter_map(|r| {
            let r = r.trim();
            if r.is_empty() {
                None
            } else {
                nova_registro::riga_letta(r)
            }
        }));
    }
    let quante = if quante == 0 { righe.len() } else { quante };
    let da = righe.len().saturating_sub(quante);
    let mut ultime: Vec<Riga> = righe.split_off(da);
    ultime.reverse();
    ultime
}

/// Oggi, in AAAA-MM-GG: quel che serve per dire «oggi» e «ieri» invece di
/// una data.
pub fn oggi() -> String {
    adesso().chars().take(10).collect()
}

/// Cerca, con le regole del Python: tutte le parole, in qualunque campo e in
/// qualunque ordine, senza accenti e senza maiuscole.
pub fn cerca(f: &Filtro, fra_quante: usize) -> Vec<Riga> {
    cerca_in(&percorso(), f, fra_quante)
}

/// Come [`cerca`], ma su un file scelto.
pub fn cerca_in(dove: &Path, f: &Filtro, fra_quante: usize) -> Vec<Riga> {
    let righe = leggi_da(dove, fra_quante);
    nova_registro::cerca(&righe, f)
        .into_iter()
        .cloned()
        .collect()
}

#[cfg(test)]
mod prove {
    use super::*;

    /// Una cartella che si cancella da sola.
    ///
    /// Le prove Rust girano **nello stesso processo**, su piu' fili: la prima
    /// stesura di queste prove spostava `APPDATA`, che e' del processo, e
    /// quattro prove su sette sono diventate rosse a caso. Qui ognuna ha il
    /// suo file e non tocca niente di condiviso.
    struct Casa(PathBuf);

    impl Casa {
        fn nuova(nome: &str) -> Casa {
            let p =
                std::env::temp_dir().join(format!("nova-registro-{nome}-{}", std::process::id()));
            let _ = std::fs::remove_dir_all(&p);
            std::fs::create_dir_all(&p).unwrap();
            Casa(p)
        }

        fn file(&self) -> PathBuf {
            self.0.join("azioni.jsonl")
        }
    }

    impl Drop for Casa {
        fn drop(&mut self) {
            let _ = std::fs::remove_dir_all(&self.0);
        }
    }

    #[test]
    fn quel_che_si_scrive_si_rilegge() {
        let casa = Casa::nuova("giro");
        annota_in(
            &casa.file(),
            "eseguito un comando",
            "C:\\lavoro",
            "Get-Process",
            "comando",
            "0",
        );
        let righe = leggi_da(&casa.file(), QUANTE);
        assert_eq!(righe.len(), 1);
        assert_eq!(righe[0].azione, "eseguito un comando");
        assert_eq!(righe[0].tipo, "comando");
        assert_eq!(righe[0].esito, "0");
        assert!(
            !righe[0].quando.is_empty(),
            "senza l'ora una riga non e' un fatto"
        );
    }

    #[test]
    fn dalla_piu_recente() {
        let casa = Casa::nuova("ordine");
        for a in ["prima", "seconda", "terza"] {
            annota_in(&casa.file(), a, "", "", "comando", "");
        }
        let righe = leggi_da(&casa.file(), QUANTE);
        assert_eq!(
            righe.iter().map(|r| r.azione.as_str()).collect::<Vec<_>>(),
            ["terza", "seconda", "prima"]
        );
    }

    #[test]
    fn una_chiave_dentro_un_comando_non_resta_sul_disco() {
        let casa = Casa::nuova("chiave");
        annota_in(
            &casa.file(),
            "eseguito un comando",
            "",
            "curl -H 'Authorization: Bearer sk-abcdefghijklmnopqrst' https://x.it",
            "comando",
            "0",
        );
        let testo = std::fs::read_to_string(casa.file()).unwrap();
        assert!(
            !testo.contains("sk-abcdefghijklmnopqrst"),
            "la chiave e' rimasta: {testo}"
        );
        assert!(testo.contains("[chiave]"));
    }

    #[test]
    fn e_nemmeno_un_valore_che_il_campo_annuncia() {
        let casa = Casa::nuova("etichetta");
        annota_in(
            &casa.file(),
            "scritto in #password",
            "https://banca.it",
            "Tramonto2026!",
            "browser",
            "",
        );
        let testo = std::fs::read_to_string(casa.file()).unwrap();
        assert!(
            !testo.contains("Tramonto2026"),
            "il valore e' rimasto: {testo}"
        );
        // La riga resta: sparisce il valore, non il fatto che sia successo.
        assert!(testo.contains("scritto in #password"));
    }

    #[test]
    fn una_riga_storta_non_ferma_le_altre() {
        let casa = Casa::nuova("storta");
        annota_in(&casa.file(), "buona", "", "", "comando", "");
        let mut f = std::fs::OpenOptions::new()
            .append(true)
            .open(casa.file())
            .unwrap();
        writeln!(f, "{{questa non e' una riga}}").unwrap();
        drop(f);
        annota_in(&casa.file(), "anche questa", "", "", "comando", "");
        let righe = leggi_da(&casa.file(), QUANTE);
        assert_eq!(
            righe.len(),
            2,
            "una riga storta si salta, non ferma la lettura"
        );
    }

    #[test]
    fn si_cerca_come_di_la() {
        let casa = Casa::nuova("cerca");
        annota_in(
            &casa.file(),
            "inviata candidatura",
            "https://lavoro.it/44",
            "Societa' Rossi",
            "browser",
            "ok",
        );
        annota_in(
            &casa.file(),
            "eseguito un comando",
            "",
            "Get-Process",
            "comando",
            "0",
        );
        let trovate = cerca_in(
            &casa.file(),
            &Filtro {
                testo: "ROSSI".into(),
                ..Default::default()
            },
            QUANTE,
        );
        assert_eq!(trovate.len(), 1);
        assert_eq!(trovate[0].azione, "inviata candidatura");
        let per_tipo = cerca_in(
            &casa.file(),
            &Filtro {
                tipo: "comando".into(),
                ..Default::default()
            },
            QUANTE,
        );
        assert_eq!(per_tipo.len(), 1);
        assert_eq!(per_tipo[0].tipo, "comando");
    }

    #[test]
    fn anche_lo_storico_si_legge() {
        // Il giorno della potatura, «cosa ho mandato tre settimane fa?» deve
        // continuare a rispondere.
        let casa = Casa::nuova("storico");
        annota_in(&casa.file(), "vecchissima", "", "", "comando", "");
        std::fs::rename(casa.file(), nova_potatura::precedente(&casa.file())).unwrap();
        annota_in(&casa.file(), "nuova", "", "", "comando", "");
        let righe = leggi_da(&casa.file(), QUANTE);
        assert_eq!(
            righe.iter().map(|r| r.azione.as_str()).collect::<Vec<_>>(),
            ["nuova", "vecchissima"]
        );
    }

    #[test]
    fn lora_e_quella_di_casa_non_quella_di_greenwich() {
        // Il fuso si passa, cosi' la prova dice la stessa cosa su una
        // macchina italiana e su quella della CI, che sta su UTC.
        assert_eq!(quando(0, 3600), "1970-01-01T01:00:00");
        assert_eq!(quando(0, 0), "1970-01-01T00:00:00");
        assert_eq!(quando(1_788_611_696, 7200), "2026-09-05T14:34:56");
        // E l'ora che si scrive davvero e' quella che il sistema dichiara.
        let t = nova_platform::orologio::adesso();
        assert_eq!(adesso(), quando(t, nova_platform::fuso_secondi(t)));
    }

    #[test]
    fn e_ha_la_forma_che_il_python_si_aspetta() {
        // `2026-08-30T14:18:05`: diciannove caratteri, niente fuso in coda.
        // Il Python la rilegge con `datetime.fromisoformat`, e una «Z» di
        // troppo la' dentro e' una riga che sparisce dal filtro per data.
        let r = adesso();
        assert_eq!(r.len(), 19, "{r}");
        assert!(r.as_bytes()[10] == b'T' && !r.ends_with('Z'), "{r}");
    }
}
